add new look for scaladoc

felixmulder opened the pull request:
I'm working on a new look for scaladoc. My goal has been to make something similar to the beautiful Typesafe sites. For now this has been for my own amusement, but if there is interest - I'd like to make this good enough to replace the current (in my opinion, somewhat outdated) look.

<img width="1440" alt="screen shot 2015-12-08 at 22 42 33" src="https://cloud.githubusercontent.com/assets/1530049/11669454/0bfab7b0-9dfd-11e5-80ce-3e45ef5ebbfe.png">
<img width="1440" alt="screen shot 2015-12-08 at 22 42 35" src="https://cloud.githubusercontent.com/assets/1530049/11669458/0bfdb320-9dfd-11e5-84e3-394e4044b306.png">
<img width="1440" alt="screen shot 2015-12-08 at 22 42 37" src="https://cloud.githubusercontent.com/assets/1530049/11669456/0bfc0a16-9dfd-11e5-89e8-d763be101410.png">
<img width="1440" alt="screen shot 2015-12-08 at 22 42 41" src="https://cloud.githubusercontent.com/assets/1530049/11669455/0bfaee7e-9dfd-11e5-8ebb-d55cdaa1e166.png">
<img width="1440" alt="screen shot 2015-12-08 at 22 42 45" src="https://cloud.githubusercontent.com/assets/1530049/11669457/0bfd7216-9dfd-11e5-9d3b-bdca27b6cf7a.png">
<img width="1440" alt="screen shot 2015-12-08 at 22 42 48" src="https://cloud.githubusercontent.com/assets/1530049/11669459/0bfe8ff2-9dfd-11e5-86b2-33c1ade60b67.png">

There's still a lot of stuff to do:
- ~~Replace large "C"-, "O"- and "T"-icons with css based images.~~
- ~~Add hide index component (left portion)~~
- Filter with: ordering, inherited, implicitly, visibility (right portion)
- ~~Add the HiDPI arrows to "Linear Supertypes/Subtypes"~~

To name a few...

One of my goals is complete removal of the static images that are not HiDPI and replacing them with utf8 symbols or CSS.

If there is interest - I'll keep working on my branch then squish for merge.

felixmulder committed:
add new look to scaladoc

This commit adds a new look to scaladoc without changing its behaviour.
Features in this commit:

- New scaladoc look
- SVG icons instead of png for HiDPI displays
- Better layout of existing features
- Better title-bar colors for each entity type: trait, object, class

Ichoran commented:
I like it personally, but my tastes aren't always representative.

hepin1989 commented:
will it looks ok on mobile?

felixmulder commented:
@Ichoran thanks - and don't sell yourself short :smile: 

@hepin1989 - I'll look into it, the current API looks "OK" on mobile, but not that great. Is there really anyone who uses the docs on mobile though?

hepin1989 commented:
@felixmulder http://elixir-lang.org/docs/stable/elixir/Kernel.html we could take a look at this:)

soc commented:
I have also looked into this ... I think it would be nicer to pick scala-lang.org as a design source. If you add the small dark blue bar at the top, and replace the existing green/blue class/trait/object pane with medium blue color, and use the light blue for the menu on the left, it could look very nicely!

felixmulder commented:
@soc I agree with you in that we should use the more "official" color scheme from scala-lang - however, I think most of the colors on the typesafe site are the same as on the official site, albeit the top portion of the main site is darker.

I'll post screenshots of your suggestion later today and we can compare.

felixmulder commented:
I've updated the first post with current screenshots as of commit 324a296.

@soc couldn't quite get the colors right - if you've got some hex codes for me I wouldn't mind trying them out.

The one thing I've left to do before this is truly compatible with the old version, is to add the "with ordering, inherited..." filter options. Added the Scala-logo as a test - should it be there? Maybe someone else doesn't want it there when they gen their docs - has this discussion been had?

cc: @VladUreche

VladUreche commented:
@felixmulder Thanks for the new design! It looks much cleaner, although maybe a bit boring due to the washed colors...

@heathermiller is the design expert, so she'll be the one who ultimately needs to approve/disprove the new design. Better coordinate her :smile:

soc commented:
@felixmulder First dark blue strip should be 20px height, #002B36. Then 80px height, #073642. Light blue is #72D0EB.

felixmulder commented:
@soc will try them out. Unfortunately the right portion of the docs is an iframe for the html file that contains the entities (package, trait et cetera). I'll figure it out somehow.

Also, the scala logo used is SVG wheareas the white scala-lang logo is png (boo). Will see if I can solve it through other means...

I'll post the next version sometime tonight.

heathermiller commented:
IT'S AMAZING!!

hepin1989 commented:
@felixmulder there is a big white place besides the scala logo,I thin you could post the current version number of the doc there.

heathermiller commented:
First off, amazing work. I can't praise you enough for how well you've visually streamlined the interface. Putting each method/member into its own distinct differently-colored box, etc... It's all really reduced the visual strain induced by the old interface. Great work.

Just a handful of thoughts:

1. Why not keep the filters collapsed by default? I find them visually noisy and I very seldom use them.
1. The Scala logo might be a bit too strong. Think of other peoples' libraries generated by Scaladoc – with such a strong "SCALA!!" shouting at you, everyones' library might appear as if its the Scala stdlib. @VladUreche do you think there'd be some way to parameterize the text/image shown in the title bar? At the very least, as a default, the name of the root directory could be used...? However, @felixmulder, in any case, it would be nice to incorporate the Scala logo somehow. It would be nice to visually mark Scala libraries as such... But then what of the people who use Scaladoc to generate Java APIs? :/
2. Would it be possible to make the search bar in the left side panel span the width of the panel? i.e., would it be possible to remove that little triangle to the left of it? Or is it functional somehow?
3. @soc's suggestion of trying to make it appear like there's one bar could be nice, I'm not sure. I have to say that his suggestion does go in the direction of further simplifying the interface, more in the direction of the Elixir docs (http://elixir-lang.org/docs/stable/elixir/Kernel.html) where clutter is absolutely minimal. But it could be that following that suggestion could mean that the sidebar is no longer perceived as a sidebar. @felixmulder, you being the designer here, I leave it to your creative discretion :)
4. Did you manage to figure out syntax highlighting or anything for `<code>` blocks? E.g., code examples that are regularly provided at the top each page.

felixmulder commented:
@heathermiller thanks!

1. The filters are collapsed by default (as in the 1st image), I also find them distracting.
2. I agree with you - as someone else suggested, if we config it by CSS people are free to change said image. It might, however, be better to choose a more discrete logo
3. The little triangle unfolds the "A B C..." as in the third image. There might be a better solution for this - I'll experiment. Letting it span the entire width would probably be prettier.
4. I wouldn't really mind having something similar to the Elixir docs, it is very nice. I'll experiment and post images regularly through the coming week.
5. I'm not quite sure I get what you mean here. But as you can see in the image below - there is already some form of syntax highlighting. Is it sufficient?

![photo279671288471070920](https://cloud.githubusercontent.com/assets/1530049/11688302/1834d45e-9e8c-11e5-8c9e-8ee667bd9194.jpg)

Note that the image above is from the first commit in this PR.

@hepin1989 where did you find the image? Is there an .svg-version? Are we allowed to use it?

heathermiller commented:
> 1. The filters are collapsed by default (as in the 1st image above), I also find them distracting.

Great. I couldn't tell if they were collapsed or just incomplete. Great to hear that they'll be collapsed by default.

> 2. I agree with you - as someone else suggested, if we config it by CSS people are free to change said image. It might, however, be better to choose a more discrete logo or none at all.

Right, in the case of someone who runs Scaladoc on a Java API, it could just be confusing. @VladUreche would you have any pointers here?

> 3. The little triangle unfolds the "A B C..." as in the third image. There might be a better solution for this - I'll experiment. Letting it span the entire width would probably be prettier.

The A-Z index is actually immensely useful. Using it, you can search for almost anything, methods, values, objects. This is counter to the search bar above it, which only searches for so-called "top-level entities", so, classes, traits, singleton objects. A great many Scala users arrive at this page and might type in something like "reduce" hoping to find documentation for some reduce method defined somewhere.

This is an artifact of a 2008-era fear of there being too many "members" (methods, fields, classes, singletons etc) for a JavaScript filter to deal with. This would actually a lot less of a concern now a days, I think. 

That said, it might be time to rethink this from a UI perspective... I know this is far beyond the scope of this PR, but if you would have any ideas for improving the searching functionality in Scaladoc from a UI perspective, they'd be welcome! If it's a nice and feasible idea, perhaps we can find a way to make it happen. Specifically, if I searched for "map" or "reduce" how should the interface behave? What happens to the left panel?

...Perhaps we can obviate the need for this A-Z index business...

> 4. I wouldn't really mind having something similar to the Elixir docs, it is very nice. I'll experiment and post images regularly through the coming week.

The sky's the limit! Scaladoc as it is was implemented to be essentially just Javadoc for Scala + Scala's features. If you think there's a better way to display documentation, we're very open to suggestions!

> 5. I'm not quite sure I get what you mean here. But as you can see in the image below - there is already some form of syntax highlighting. Is it sufficient?

That's great, I didn't see it posted. Looks good!

lrytz commented:
> The A-Z index is actually immensely useful. Using it, you can search for almost anything, methods, values, objects.

I wasn't even aware..!

A search box for all 

> This is counter to the search bar above it, which only searches for so-called "top-level entities", so, classes, traits, singleton objects. A great many Scala users arrive at this page and might type in something like "reduce" hoping to find documentation for some reduce method defined somewhere.
> 
> This is an artifact of a 2008-era fear of there being too many "members" (methods, fields, classes, singletons etc) for a JavaScript filter to deal with. This would actually a lot less of a concern now a days, I think.
> 
> That said, it might be time to rethink this from a UI perspective... I know this is far beyond the scope of this PR, but if you would have any ideas for improving the searching functionality in Scaladoc from a UI perspective, they'd be welcome! If it's a nice and feasible idea, perhaps we can find a way to make it happen. Specifically, if I searched for "map" or "reduce" how should the interface behave? What happens to the left panel?
> 
> ...Perhaps we can obviate the need for this A-Z index business...

lrytz commented:
ups, sorry..

felixmulder commented:
@heathermiller like @lrytz I too was not aware of what it did until I started fidgeting with a redesign.

The searchbox should definitely search for everything - that would be the expected usage. It's a bit outside the scope; we should, however, aim to keep it in mind when designing this interface.

heathermiller commented:
> The searchbox should definitely search for everything - that would be the expected usage. It's a bit outside the scope; we should, however, aim to keep it in mind when designing this interface.

Sure. But if you have an idea/sketch of how you think it should behave, it's a starting point for figuring out how to make it happen.

lrytz commented:
:+1:

felixmulder commented:
I simplified the sidebar to be more minimal.

<img width="1440" alt="screen shot 2015-12-09 at 20 55 40" src="https://cloud.githubusercontent.com/assets/1530049/11697073/245356d0-9eb8-11e5-97da-0f75b15919af.png">
<img width="1440" alt="screen shot 2015-12-09 at 20 55 52" src="https://cloud.githubusercontent.com/assets/1530049/11697072/244f70ba-9eb8-11e5-8edd-937d1571f174.png">
<img width="1440" alt="screen shot 2015-12-09 at 20 56 03" src="https://cloud.githubusercontent.com/assets/1530049/11697071/244e4938-9eb8-11e5-8493-c35bb95cee46.png">

It retains the "..." as a means to get to the A-Z functionality.

I'm no designer - I study compiler technology and algorithms. My brother on the other hand is an actual designer. He came up with this for the folded up buttons at the top of the class:

![812740526_15601411063764634077](https://cloud.githubusercontent.com/assets/1530049/11697174/acb8fa98-9eb8-11e5-84d4-3b43633b1114.jpg)

I think it looks pretty neat and ditches the skeuomorphism from the previous edition. 

I tried letting the top cover the entire screen horizontally, it looks weird to me. I like the idea, but it doesn't feel like there's a good partition between the class description and the "bar" - what do you think? What I mean:

<img width="1440" alt="screen shot 2015-12-09 at 21 14 31" src="https://cloud.githubusercontent.com/assets/1530049/11697452/f1dd3d18-9eb9-11e5-8930-44657282647e.png">

cc: @heathermiller

SethTisue commented:
1) This is awesome.
(but while I've got your attention, I can't resist bringing one issue up:)
2) The number-one usability problem with Scaladoc for beginners, in my experience, is the non-discoverability of the toggle between the class and the companion object. Newbies don't realize it exists. So anything you can do to make that a great more obvious would be a good.

som-snytt commented:
TIL I'm basically a skew-o-morph. My kid might be one, too.

A big red scream balloon, "Check under the sticker to win amazing new free gear!", might suffice to draw attention to the companion button.

felixmulder commented:
@SethTisue yeah, you're totally right. I'll think about how to emphasize its clickability. As of now this is the current look:

<img width="1440" alt="screen shot 2015-12-10 at 09 51 25" src="https://cloud.githubusercontent.com/assets/1530049/11711046/02a16588-9f24-11e5-98c1-a004fb23c547.png">

heathermiller commented:
@SethTisue didn't Dick Wall at least partially resolve this issue by putting the big link to companion on the right side of the class/object name under "Related Docs"? 

@felixmulder Back to the issue of the index... I strongly advocate for **not** hiding it. It is already bad that a lot of people don't realize that it's there. It's worse if we make the index harder to access for those who regularly use it. Granted, I take this stance of not hiding the index assuming we're not going to have a conversation about how we can improve the search/filtering experience in Scaladoc. I'm still happy to brainstorm together to try and imagine ways that this experience can be improved.

About the C/O etc images at the top... I like them! The class image though, is it me or is the blue behind the green circle a bit bright?

felixmulder commented:
@heathermiller 
> About the C/O etc images at the top... I like them! The class image though, is it me or is the blue behind
> the green circle a bit bright?

Yeah, that one's a mockup - the actual one is darker. 

> I strongly advocate for **not** hiding it

Perhaps we could have the letters vertically on the right side of the bar? (Visible always)

soc commented:
Maybe we could show all available kinds (object/trait/...) at the top, just like in the index on the left?

felixmulder commented:
@soc I'm concerned it might look cluttered - but I'll mock it up after I get home from thesis work. Another option would be to add some form of (tab-)buttons on the right instead of the "Related docs" hrefs?

felixmulder commented:
For anyone looking at the failing build - I've removed some assets (png-files) that the HtmlFactory is copying around, will fix in next commit.

soc commented:
@felixmulder I think you could only show the outlines of the ones not selected.

I. e. invert the colors of everything in the circle in https://cloud.githubusercontent.com/assets/1530049/11697764/8158053a-9ebb-11e5-9512-e04f801b0934.png

SethTisue commented:
> didn't Dick Wall at least partially resolve this issue by putting the big link to companion on the right side of the class/object name under "Related Docs"

good point. but yeah, "partially" — it's still very easy to miss.

felixmulder commented:
I had an issue with `HtmlFactory.scala`. When adding assets to the `libResources` list, it couldn't find them despite them being in the correct folder. As such I've (for now) commented out the new assets. The fonts should not be able to be copied via the current `copyResource(subPath: String)` function though, so I'll need to remedy that before this will pass any tests.

Still todo:

* Related docs 
* Filter with: ordering, inherited, implicitly, visibility (right portion)

Other than that, I think it's nearing completion...

heathermiller commented:
Again, I would really advocate that the index is not hidden.

SethTisue commented:
> I would really advocate that the index is not hidden.

I agree with Heather on this.

heathermiller commented:
Maybe you could leave the "display only packages" button collapsed, and keep the A-Z index where it is? Just brainstorming...

felixmulder commented:
<img width="446" alt="screen shot 2015-12-11 at 19 19 27" src="https://cloud.githubusercontent.com/assets/1530049/11751815/2fbffbec-a03c-11e5-9701-9b3b00083652.png">

How's this as an alternative?

I think we should re-implement search. If we do that, we should get rid of the index, it's confusing. When I click on the letters I would think that the bar would jump to classes beginning with the letter clicked... :confused:

heathermiller commented:
Yep, I totally agree that we should re-implement search. I have an idea of how we can change Scaladoc to generating a different index (I wouldn't expect you to do it, don't worry). I just don't know how would be best to visually display the search results in a meaningful way.

Think about it, if you search for "bool" you should see a ton of things, the Boolean class/companion object, and then all of the methods and fields that contain the word boolean in them. That said, naively, I would guess that the interface should separate out the classes/traits/objects (so-called "entities") from methods/fields/etc (so-called "members"). So maybe they should be separated by something like a header that says "classes/traits/objects" and those should sit on top of the filtered results, and then maybe the second category "methods, fields, members" or something could be another header that sits below that list.

Do you see what I mean?

heathermiller commented:
Here's an example of information that I'm not sure how we'd display:
![screen shot 2015-12-11 at 8 02 53 pm](https://cloud.githubusercontent.com/assets/687163/11752873/41364ff6-a042-11e5-9bd3-7c3a41245007.png)

Here are "members" that are called "bool" something but they live on different classes... For example, "Boolean" in the scala package object, as a member on ClassTag, etc. See what I mean?

heathermiller commented:
Btw, +1 on the switch thingy :)

felixmulder commented:
Yes! That's exactly the way I want it to work! I don't want to implement it in this PR, but I wouldn't mind doing it at a later point before the 2.12.x-RCs...let's keep discussing it.

heathermiller commented:
Sure thing – I'd be happy to spend some time working on it too. Maybe @VladUreche would also be available one of these days to do some brainstorming on generating the index.

But for the meantime, can we just keep the:
A B C D E F G H I J K L M N O P Q R S T U V X Y Z # thing?

Rationale: even if we do "A-Z" it's still the same as having to click to expand. It adds a click to power users' workflow.

heathermiller commented:
Ah – and one last (stupid) question. Are the keyboard shortcuts still good?
There are only like 3-4: http://heather.miller.am/scaladoc-keyboard-shortcuts.html

felixmulder commented:
They should be working. How about a cookie for remembering if the letters were showing?

Otherwise at the very least I'll make it look better, hopefully :smiley:

felixmulder commented:
I'm now done with the first final draft of this PR - please try building before merging. No functionality has been changed as previously discussed. The resulting build should generate a doc that looks like this:

<img width="1440" alt="screen shot 2015-12-13 at 14 28 30" src="https://cloud.githubusercontent.com/assets/1530049/11767676/8a87173a-a1b4-11e5-9a77-f429930d2fd2.png">

It has been tested on OS X in Chrome and Firefox.

felixmulder commented:
There's one known issue - missing an SVG for trait with companion object. Will add this tonight - but it doesn't change anything from the final draft otherwise :smile:

felixmulder commented:
Fixes have been squished into c91fdb0. These fixes include

- SVG for trait companion icon
- Proper color title-bar color for trait

heathermiller commented:
Great! This looks really nice. One last request before merging – would you mind generating the stdlib API docs and dropping them in a publicly accessible directory somewhere? Would be great to be able to try this out before merging it.

felixmulder commented:
Sure thing, I'll think of someplace to host them. (Unfortunately, LTH does not permit traffic from external IPs to the student public dirs.)

I'll find someplace to host them and then ping you - should be sometime tonight.

lrytz commented:
Dropbox should work: copy the generated docs into a folder below Public and get the public link to `index.html`

felixmulder commented:
@lrytz thanks!

felixmulder commented:
I added an underline to the search bar on the right pane. As such - Jenkins is re-testing everything.

Nevertheless, here's the current build:

https://dl.dropboxusercontent.com/u/358427/scaladoc/index.html

@heathermiller

heathermiller commented:
Awesome, thank you very much! It looks amazing. Improvement of 10,000x over before. I just noticed the permalink icon! Couldn't see it in the screenshots. It's awesome!!

Two smallish bugs that I noticed... 
1. Keyboard shortcuts. When I type something in the left side panel, and press the down arrow key, old CSS seems to be used for the selected trait/class/object – the height and there appears to be an old png bg still in place. Should be a quick fix.
2. When I load a page via permalink, old CSS seems to be used – there's a gaudy orange highlight. I think it might actually be js. The scroll is also a bit off for the members I clicked (i.e., due to the top bar now being anchored, I think the calculation is now off for the js scrolling if you know what I mean). Do you think it'd be possible to highlight the bg of the permalinked member a different color?

heathermiller commented:
Ah, a new little bug... When I mouse over the top bar (anywhere), a permalink icon appears in the top right hand corner of the page. For the class/trait/object/top bar thing, it of course doesn't do anything. So this should be removed.

heathermiller commented:
Ok, other than that, this all looks GREAT to me! Do you think those 3 little fixes above are manageable?

felixmulder commented:
Will see to it when I get home from praktice. What browser and OS are you using?

soc commented:
This is great!

A few issues I found though:

- The JavaScript seems to be extremely slow (Firefox asks whether it should stop running JavaScript on that site when loading anything)
 - Imho, "Filter functions" should not be white, makes it hard to figure out that one can click into it. Maybe something in gray, or maybe something visually similar to the filter input box?
- Fonts in the icons in the index seem to be off
- Upper part of the scrollbar seems to be hidden by the filter box in the upper left

heathermiller commented:
Hi @felixmulder, thanks! OSX + Chrome. Really excited about this overhaul!!

(PS, I'm aware that Dropbox is remarkably slow at serving sites...)

felixmulder commented:
@soc 

> The JavaScript seems to be extremely slow (Firefox asks whether it should stop running JavaScript on > that site when loading anything)

I haven't changed anything that should impact performance, can you identify what the bottleneck is? Perhaps it's loading a large entry like the Array class?

> Imho, "Filter functions" should not be white, makes it hard to figure out that one can click into it. Maybe > something in gray, or maybe something visually similar to the filter input box?

I don't think it's the "white" that is the issue, but rather the fact that it doesn't look like an input field, I might darken the background in the input box or something. Good feedback anyhoo!

> Fonts in the icons in the index seem to be off

Looks like it is specific to firefox - I'm gonne see what I can do about it.

> Upper part of the scrollbar seems to be hidden by the filter box in the upper left

On it.

PS - it would really help if you guys note what OS you're using.

felixmulder commented:
> 1. Keyboard shortcuts. When I type something in the left side panel, and press the down arrow key, old CSS seems to be used for the selected trait/class/object – the old (shorter) height of the bar seems to be used, which causes resizing when scrolling around with the keyboard, and there appears to be an old png bg still in place. Should be a quick fix.

Will fix, missed that one.

> 2. When I load a page via permalink, old CSS seems to be used – there's a gaudy orange highlight. I think it might actually be js. The scroll is also a bit off for the members I clicked (i.e., due to the top bar now being anchored, I think the calculation is now off for the js scrolling if you know what I mean). Do you think it'd be possible to highlight the bg of the permalinked member a different color?

Can't reproduce - give me a permalink?

> 3. When I mouse over the top bar (anywhere), a permalink icon appears in the top right hand corner of the page. For the class/trait/object/top bar thing, it of course doesn't do anything. So this should be removed.

Clicking it changes the URL in the address-field to the permalink - or am I missing something?

heathermiller commented:
Re #2, here's a link:
https://dl.dropboxusercontent.com/u/358427/scaladoc/index.html#scala.Option@reduceOption[A1>:A](op:(A1,A1)=>A1):Option[A1]

felixmulder commented:
Allright - I applied fixes for the following:

* fonts in index icons being off in firefox
* upper part of scrollbar hidden in firefox
* keyboard nav highlighting
* fix permalink highlighting of functions in entity

Things I haven't fixed:

* @heathermiller's 3rd issue - no understandy 
* "Filter functions" input not looking editable

The last two I'll look at tomorrow if applicable :smile:

soc commented:
This is how things look when I click the "debug script" in the "warning: unresponsive script" dialog: http://i.imgur.com/lUZ8CK7.png

This has never occurred with the old one, but Array hangs every time.
Additionally, there are lots of duplicate entries:

http://i.imgur.com/pyaqmyL.png

soc commented:
Also: A lot of "not so helpful" classes have appeared in the index, like "CollectionsHaveToParArray". Was this intentional?

soc commented:
The bar that appears after clikcing "focus" hides parts of the elements.

ctongfei commented:
This is fabulous!
But it seems that I cannot build it correctly. I checkout the branch and called the main class scala.nsc.tools.ScalaDoc but the generated ScalaDoc is still in the original style. Can anyone tell me what is the correct way to run this?

SethTisue commented:
@ctongfei from a clean checkout of Felix's branch, `ant docs` should be sufficient — the generated doc for the standard library will end up in `build/scaladoc/library`. and if you want to run scaladoc from the command line, after `ant quick.bin` you can run `build/quick/scaladoc`. note you can get help with this kind of thing fastest in https://gitter.im/scala/contributors

nafg commented:
Suggestion for searching members: When you start typing in the top-level search box, there could be a link that the top that says something like "search all members..." that when clicked takes you to a search page. (Without getting into how to implement that page.)

By the way, what exactly is included in the default search / in the sidebar? Is it "members of a package" (including defs in a package object)? Is it "objects, classes, and traits" (including inner objects, classes, and traits)? Is it "things that are the top level of the source file" (that would exclude classes within a `package { ... }` block)?

felixmulder commented:
@soc 

> This is how things look when I click the "debug script" in the "warning: unresponsive script" dialog: http://i.imgur.com/lUZ8CK7.png

Yeah that's the jQuery script - i.e. one that hasn't been modified.

> This has never occurred with the old one, but Array hangs every time.
> Additionally, there are lots of duplicate entries:
> 
> http://i.imgur.com/pyaqmyL.png

The "duplicate" entries also appear in the upstream of the scala repository - however, not in the nightly build. Can someone answer why this is?

<img width="1440" alt="screen shot 2015-12-15 at 07 12 18" src="https://cloud.githubusercontent.com/assets/1530049/11803447/4cf02630-a2fb-11e5-981a-fba1e6892c0c.png">
<img width="1440" alt="screen shot 2015-12-15 at 07 12 23" src="https://cloud.githubusercontent.com/assets/1530049/11803448/4cf4a282-a2fb-11e5-9eee-8e08d3308443.png">

> Also: A lot of "not so helpful" classes have appeared in the index, like "CollectionsHaveToParArray". Was this intentional?

This class also appears in the upstream version - so someone else is gonna have to take the heat for that one.

> The bar that appears after clikcing "focus" hides parts of the elements.

Thanks, will get right on this one!

felixmulder commented:
> Suggestion for searching members: When you start typing in the top-level search box, there could be a link that the top that says something like "search all members..." that when clicked takes you to a search page. (Without getting into how to implement that page.)

Great suggestion! This is something along the lines of what we're looking to implement. But it will be done in a different PR :)

> By the way, what exactly is included in the default search / in the sidebar? Is it "members of a package" (including defs in a package object)? Is it "objects, classes, and traits" (including inner objects, classes, and traits)? Is it "things that are the top level of the source file" (that would exclude classes within a package { ... } block)?

See @heathermiller's comment about "top level entities"

felixmulder commented:
@soc's and @heathermiller's input from yesterday  evening is now live here: https://dl.dropboxusercontent.com/u/358427/scaladoc/index.html#scala.Boolean

It's still uploading, so some entities might give you a dropbox-404.

hepin1989 commented:
I am using chrome version 47,and
![image](https://cloud.githubusercontent.com/assets/501740/11830708/80e171be-a3e2-11e5-97bf-d80f85469d82.png)
and you see,it clasp with the scrollbar.but after click it's ok
![image](https://cloud.githubusercontent.com/assets/501740/11830761/e3a81e06-a3e2-11e5-9a97-2786b27ced62.png)


how about make the  `foldAll` be a switch and align side with the `deprecated`?

felixmulder commented:
@hepin1989 that's one alternative, it might be messy to detect the scrollbar and move the button if it exists...might be better to place it up top somewhere. I'll look at it tonight!

felixmulder commented:
@soc are the letters in the index centered for you now?

https://dl.dropboxusercontent.com/u/358427/scaladoc/index.html#scala.Some

heathermiller commented:
Issue #2 that I mentioned above is still present :/

felixmulder commented:
@heathermiller fixed?

heathermiller commented:
Sort of. Yes on the scrolling issue on page load (thank you!), but there's still a scrolling issue when you click a permalink on an already-loaded page. Two tiny tweaks:

1. Issue: when I click the permalink (without reloading the page) it scrolls up and is under the top bar. 
2. When a permalink is clicked (or reloaded) rather than this flash/fade out, can we just have the background change to a solid color so it's easy to see which member is permalinked to? E.g., why not reuse #c2d2dc? We already use it as a way to highlight members :)

Here's a gif to show what I mean... First I click the permalink to show the scroller scrolling too high on an already-loaded page, and then I load a page from the permalink (here it works, but the fade is too light, I don't visually find the member I'm permalinked to, a more distinct color would help here).

![screencast4](https://cloud.githubusercontent.com/assets/687163/11839020/ff97a520-a3ea-11e5-9bdb-fdcee92ef45b.gif)

felixmulder commented:
@heathermiller: thanks for the help, think I fixed it now - please verify on the dropbox.

felixmulder commented:
@heathermiller fixed

heathermiller commented:
AMAZING! It looks great!!

I gotta say, this is really perfect! Such an awesome job @felixmulder – THANK YOUx10^6!!!

LGTM :)

felixmulder commented:
Allright - so we've gotta fix two things before this can be merged (if there are no more issues)

- Making "Filter functions" look editable
- "Fixing" the offset of the "Fold All" button on browsers with scrollbars (i.e. non OS X)
- Squishing the history

Any further input?

heathermiller commented:
For "making filter functions look editable" what about something like this?

![fullsizerender](https://cloud.githubusercontent.com/assets/687163/11840820/228939fa-a3f9-11e5-9a1c-e86dfe0db636.jpg)

heathermiller commented:
Or we could also do something like the permalink, where an icon appears on hover over? Slack has a similar thing, for example:

![screencast5](https://cloud.githubusercontent.com/assets/687163/11840900/d0d40c92-a3f9-11e5-8e6a-59b25b793d69.gif)

felixmulder commented:
I like it. Sounds like we can combine the two, I'll get back to you once I've done something about the two remaining issues.

felixmulder commented:
Confirmed that 9554c1e also works on OS X (retina). One last (known) issue to go! Woop woop!

felixmulder commented:
I've fixed everything I could think of and posted the changes (not uploading the entire thing again) on dropbox: https://dl.dropboxusercontent.com/u/358427/scaladoc/index.html#scala.Boolean

If there are no more issues - I feel done with this PR @heathermiller

hepin1989 commented:
@felixmulder thank you for the great work,the new version I could not find the `Type Hierarchy Diagram`,is that by design? 
![image](https://cloud.githubusercontent.com/assets/501740/11868944/8c5203e6-a4f6-11e5-849f-4db2e653d580.png)

felixmulder commented:
@hepin1989 it hasn't been removed - but when generating the docs with `ant docs` the script did complain about the diagram failing to generate. This has nothing to do with the changes, however, it might be good to generate them to see that they are compatible with the new theme.

I'll look into it ASAP.

hepin1989 commented:
@felixmulder thanks ,and how about hide this `x` when the filter input is empty ?
![image](https://cloud.githubusercontent.com/assets/501740/11869200/b4b96692-a4f8-11e5-81f5-6748ffd1c79f.png)

and how about add the same `filter` apperance to the top left one?
![image](https://cloud.githubusercontent.com/assets/501740/11869227/fe6a0602-a4f8-11e5-99ac-122b758747ce.png)

another improvement I think how about add directly jump from the `method` to the method defined in the source code?

like `</>` tag in elixir doc

![image](https://cloud.githubusercontent.com/assets/501740/11869271/68545680-a4f9-11e5-8c7b-385822a31604.png)

when you click it ,it directly jump to the method definition in the source code. I think that should be very very helpful.

felixmulder commented:
@hepin1989 added your suggestion regarding the "x".

I will not, however add the direct jump thing - it is a good suggestion. But for now, and for this PR, I'm satisfied with having the source available at the top of the page. Plus, we'd have to rethink the permalink thing.

Maybe this is something you could do in a later PR?

heathermiller commented:
@felixmulder It looks good! One problem though, in my sketch above, I renamed "Filter functions" to "Filter members". "Functions" isn't quite correct terminology here since a member can be a type alias, a field, a method, etc. So can you please update it so as to correct the terminology?

Otherwise this LGTM.

heathermiller commented:
You could also write "Filter for fields, methods, type aliases, etc" if you feel it is more clear. I just object a bit to the use of "functions" here.

felixmulder commented:
@heathermiller how about "Filter members, values etc" ?

hepin1989 commented:
@felixmulder the `permalink ` in elixir's doc is on the left side,and only appear once your mouse hover.I think this PR is bigger enough just as you said.

The direct jump link is a good addition too,but maybe a separate ticket then.

heathermiller commented:
Well, a value is a _kind_ of member. So I'd either say just "members" or I'd list what a member is, e.g., "methods, values, type aliases, etc" rather than mixing the definition with the term itself if you know what I mean

SethTisue commented:
"Filter members" seems accurate to me

felixmulder commented:
@SethTisue @heathermiller "Filter members" it is :+1:

felixmulder commented:
Fixed the placeholder - ready for merge @heathermiller

heathermiller commented:
Woohoo! LGTM!! THIS IS AMAZING! Thank you @felixmulder!!

SethTisue commented:
I've asked the compiler team if they have any issues with merging this (now that all the activity has settled down)

heathermiller commented:
The changes are only template HTML/CS/JS so I would hope that they don't take issue...

ctongfei commented:
This is purely amazing.
However, the font face is sometimes inconsistent: For monospace, sometimes it is a webfont "Source Code Pro", and sometimes it's the default value "monospace" which will be rendered as Consolas or Menlo. There's also the problem with "Lato" vs "sans".

felixmulder commented:
re @ctongfei:

- The Lato typeface is used for the left bar entities only, this is as intended. Open Sans is used for bread text.
- I want to clarify one thing regarding the "webfont": the "Source Code Pro" typeface you are referring to is an open source typeface that is freely available and included in the project as an asset. Specifically I want to make clear that there are no loading of remote resources in the docs. There might be some confusion here since Google does host the font in their font CDN.
- Source Code pro is used in the entity description "&lt;pre&gt;" (i.e. code snippets) sections exclusively. This is because  it looks horrendous when used inline, IMO.

If you can find places where Source Code Pro should be used, but isn't - please let me know.

Regarding the "problem" with Lato/Sans, you're going to have to be more specific, I don't understand what you mean.

felixmulder commented:
I think the changes in this commit fix SI-9598

felixmulder commented:
Also seems to remedy SI-9597

cc @janekdb

janekdb commented:
@felixmulder - Thanks. Comments added in Jira.

I am seeing a slow down on the "Fold All"/"Unfold All" behaviour. For the first seven clicks it's fast enough. On the eighth click I can sense a slow. By the time I get to the 15th click I've counted off 30 seconds and Chrome has offered to kill the slow running script.

janekdb commented:
Apologies if I've not read a possible caveat about https://dl.dropboxusercontent.com/u/358427/scaladoc/index.html and what to expect of it. If I assume it's meant to be a fully working site then group headers will need some work.

Group Headers Missing:
https://dl.dropboxusercontent.com/u/358427/scaladoc/index.html#scala.language$|

Group Headers Present:
http://www.scala-lang.org/api/current/#scala.language$|

felixmulder commented:
@janekdb confirmed the slowness of clicking fold/unfold - and fixed it. I'll push the fix momentarily.

Regarding the missing file - it is probably missing because my connection was poor when uploading the docs so I only uploaded about 80% of the files from the standard lib. I'll check on it!

felixmulder commented:
Fix for the sluggish fold/unfold conundrum has been pushed in 94e7cce and is now live on: dl.dropboxusercontent.com/u/358427/scaladoc/index.html

felixmulder commented:
@janekdb regarding Group Headers, I'm not sure what you mean. Both URLs you have supplied lead me to 404-sites. The dropbox one gives a dropbox-404 and the scala-lang one gives me a blank page saying "Not Found".

felixmulder commented:
<img width="1440" alt="screen shot 2015-12-21 at 23 24 27" src="https://cloud.githubusercontent.com/assets/1530049/11942897/2e6a7050-a83a-11e5-83e5-0fc3b3fe7ba2.png">
<img width="1440" alt="screen shot 2015-12-21 at 23 24 37" src="https://cloud.githubusercontent.com/assets/1530049/11942898/2e6b6488-a83a-11e5-8c99-db3a90cecf1f.png">

Thanks for the input @janekdb - `@groupdesc` is now styled according to the new look.

felixmulder commented:
The complete docs are now available on dropbox. Since Janek found something in the reflect library on a hunch, I think it's best if the entire build is made readily available.

https://dl.dropboxusercontent.com/u/358427/scaladoc/library/index.html
https://dl.dropboxusercontent.com/u/358427/scaladoc/reflect/index.html
https://dl.dropboxusercontent.com/u/358427/scaladoc/scalap/index.html
https://dl.dropboxusercontent.com/u/358427/scaladoc/compiler/index.html

nafg commented:
FWIW it still says "Filter functions"

On Mon, Dec 21, 2015 at 5:53 PM Felix Mulder <notifications@github.com>
wrote:

> The complete docs are now available on dropbox. Since Janek found
> something in the reflect library on a hunch, I think it's best if the
> entire build is made readily available.
>
> https://dl.dropboxusercontent.com/u/358427/scaladoc/library/index.html
> https://dl.dropboxusercontent.com/u/358427/scaladoc/reflect/index.html
> https://dl.dropboxusercontent.com/u/358427/scaladoc/scalap/index.html
> https://dl.dropboxusercontent.com/u/358427/scaladoc/compiler/index.html
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scala/scala/pull/4877#issuecomment-166445908>.
>

VladUreche commented:
@felixmulder sorry for chiming in so late, I've been flying under the radar to get some academic work done. The new layout looks awesome, and the colors are perfect! Thanks a lot for the extraordinary work you pulled off!

There's are two little snags that I still see:

**(1)** The missing "abstract type" icon: https://dl.dropboxusercontent.com/u/358427/scaladoc/reflect/index.html#scala.reflect.macros.Universe$Run

So, long story short, reflection uses a virtual class-like pattern:
```scala
trait ReflectApi {
  type Symbol <: SymbolApi
  val Symbol: SymbolExtractor
  val SymbolClassTag: ClassTag[Symbol]
}

trait ReflectImplementation {
  type Symbol = // SymbolImpl (the implementation)
  val Symbol = // SymbolImpl companion object
  val SymbolClassTag: ClassTag[Symbol] = classTag[SymbolImpl].asInstanceOf
}
```
The problem is that `ReflectImplementation` is not visible to users, so unless we generate pages for abstract types such as `type Symbol`, there's no API visible. This is how we end up with a page for `type ClassSymbol`: https://dl.dropboxusercontent.com/u/358427/scaladoc/reflect/index.html#scala.reflect.api.Symbols$ClassSymbol

But it seems that currently these don't get a proper image at the top (and I'd also like to double-check that they appear correctly in diagrams). For more information on the annotations that produce these pages, see [this presentation](http://fr.slideshare.net/VladUreche/scaladoc-reflection).

Maybe an easy test case is:
```scala
trait T {
  /** Type Boo.
   * @template
   */
  type Boo <: String
}
```

**(2)** The groups are shown very nicely but does it look as well if the group has a description? You can try it with:
```scala
/** 
 * Class C
 * @groupdesc foo "The foo-like methods in [[C class C]]:
 * @groupdesc bar "The bar-like methods in [[C class C]]:
 */
class C {
  /**
   * Method foo
   * @group foo
   */
  def foo = ???

  /**
   * Method bar
   * @group bar
   */
  def bar = ???
}
```
Below the "foo" group, you should have the "The foo-like methods ..." text.

Thanks!

felixmulder commented:
@nafg - I was too lazy to regenerate the entire docs... so it _was_ fixed all along. I have now regenerated the docs and upped them to Dropbox. Should show you "Filter members" now instead.

@VladUreche awesome feedback Vlad!

1. I have added the abstract type icon: https://dl.dropboxusercontent.com/u/358427/scaladoc/reflect/index.html#scala.reflect.api.Symbols$ClassSymbol
2. The example you suggested for groups: https://dl.dropboxusercontent.com/u/358427/scaladoc/reflect/index.html#scala.reflect.api.C

Regarding 2, I found a real-world example here: https://dl.dropboxusercontent.com/u/358427/scaladoc/reflect/index.html#scala.reflect.api.Annotations

When it comes to the type hierarchies, I managed to get graphviz working as can be seen in the "Annotations" of the reflect API (URL above). As you can see, it looks the same as before. I think for this PR, it is good enough. I can fix it in a separate PR, after this one is merged. I don't, however, feel that I want to invest time into it if this won't be merged for some reason.

VladUreche commented:
@felixmulder Thanks for this last batch of changes! I don't think there's any reason this patch wouldn't be merged, it's an awesome addition to Scaladoc!

Ping @scala/community-maintainers: speak now or forever hold your peace.

adriaanm commented:
Looks fantastic to me -- thanks everyone!

VladUreche commented:
@felixmulder, can you please rebase on `2.12.x`? I can't merge otherwise :cry:. 

![screenshot from 2015-12-22 19 08 18](https://cloud.githubusercontent.com/assets/428706/11962338/6425b402-a8df-11e5-9c0d-ddfdb0cf3725.png)

felixmulder commented:
@VladUreche done :beers:

VladUreche commented:
Great! Once all tests pass, I'm merging.

VladUreche commented:
@felixmulder thank you for such a great contribution!

felixmulder commented:
Pop the champagne!

Would like to thank everyone involved in this PR by either suggesting improvements or helping rid the new docs of bugs. A special thanks to: @heathermiller @VladUreche @janekdb @soc @hepin1989 @SethTisue 

Excited to get to use this :)

heathermiller commented:
WOOOO!! :clap: :beers: :tada: :+1:

SethTisue commented:
We'll mention this prominently in the 2.12.0-M4 release notes.

VladUreche commented:
Btw, @SethTisue do you have any idea why the new look is not showing up in the [nightly](http://www.scala-lang.org/files/archive/nightly/2.12.x/api/2.12.x/?_ga=1.216158294.1519950162.1375191495#package) [builds](http://www.scala-lang.org/files/archive/nightly/2.12.x/api/2.12.x/scala-compiler/?_ga=1.216158294.1519950162.1375191495#package)?

SethTisue commented:
> @SethTisue do you have any idea why the new look is not showing up in the nightly builds?

@VladUreche thanks for calling this to our attention. I've opened https://github.com/scala/scala-dev/issues/70 on it. we'll fix, but perhaps not before January.

MasseGuillaume commented:
Some feedback from Reddit: https://www.reddit.com/r/scala/comments/3yeskn/new_scaladoc_design/cycudb2

VladUreche commented:
@MasseGuillaume Thanks!

mobiworx commented:
very, very cool!!! how about icons instead of "hide" and "focus" ?

muuki88 commented:
Awesome work. Sorry if this question has already been asked, but will this be backported to 2.11.x?

VladUreche commented:
@muuki88 Not sure, but I'll let @SethTisue comment on that

One more issue, found in https://dl.dropboxusercontent.com/u/358427/scaladoc/reflect/index.html#scala.reflect.api.Universe: here's how the code box looks on my machine (Ubuntu 12.04 with Firefox 43.0):

![screenshot from 2015-12-30 11 38 50](https://cloud.githubusercontent.com/assets/428706/12049656/2b22286e-aeea-11e5-8e35-378f873a2260.png)

The bottom scroll bar is redundant and shouldn't be shown.

SethTisue commented:
> will this be backported to 2.11.x?

(not speaking for the whole Scala team on this, just my own take:)

We only have [two more 2.11.x releases planned](http://www.scala-lang.org/news/2016-schedule/), one of which is imminent (and already late). And we don't do RCs on the 2.11.x series. Under the circumstances, I'd be very uneasy about accepting a 2.11.x backport, since if problems are found, it might be a long time before fixes could be released.

Once the new code has had many eyes on it and been proven and found stable in the 2.12 context, I think we could eventually reconsider.

VladUreche commented:
@SethTisue In you decide to backport, please wait until #4888 is also merged, as it improves the readability on mobile devices!

adriaanm commented:
> Once the new code has had many eyes on it and been proven and found stable in the 2.12 context, I think we could eventually reconsider.
+1

heathermiller commented:
For the record, I think that @adriaanm and @SethTisue's sentiment is far too overly cautious in the context of a pure HTML/CSS facelift. I would agree with this sentiment for a change to actual stdlib/Scaladoc code. But I think it's a hands-down improvement to the readability and usability of Scaladoc. It makes the experience of viewing documentation so much markedly better that I can't see any sane reason to force it to sit on the 2.12 backburner for another 8 months (optimistic) to a year (likely).

nafg commented:
Is there any way to provide a fallback? Or to enable people to use it like
an add-on?

On Wed, Dec 30, 2015, 6:12 PM Heather Miller <notifications@github.com>
wrote:

> For the record, I think that @adriaanm <https://github.com/adriaanm> and
> @SethTisue <https://github.com/SethTisue>'s sentiment is far too overly
> cautious in the context of a pure HTML/CSS facelift. I would agree with
> this sentiment for a change to actual stdlib/Scaladoc code. But I think
> it's a hands-down improvement to the readability and usability of Scaladoc.
> It makes the experience of viewing documentation so much markedly better
> that I can't see any sane reason to force it to sit on the 2.12 backburner
> for another 8 months (optimistic) to a year (likely).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scala/scala/pull/4877#issuecomment-168092522>.
>

felixmulder commented:
@nafg: Subclassing "HtmlFactory" and changing the assets copied should do the trick. There'd also need to be some parameter introduced to scaladoc to choose the subclass. Might be oversimplifying/uglyfing it, but in essence that would work.

nafg commented:
It would be cool if it could be available in 2.11.8 with a way to switch in
case there are bugs

On Mon, Jan 4, 2016, 1:59 AM Felix Mulder <notifications@github.com> wrote:

> @nafg <https://github.com/nafg>: Subclassing "HtmlFactory" and changing
> the assets copied should do the trick. There'd also need to be some
> parameter introduced to scaladoc to choose the subclass. Might be
> oversimplifying/uglyfing it, but in essence that would work.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scala/scala/pull/4877#issuecomment-168598405>.
>

hepin1989 commented:
@felixmulder how about add this ?
![image](https://cloud.githubusercontent.com/assets/501740/12439069/b2f18c98-bf67-11e5-947f-62fbdf69b670.png)
I cut from here
https://hackage.haskell.org/package/Stream-0.4.7.2/docs/Data-Stream.html

nafg commented:
@hepin1989 don't you think that's a bit more than aesthetics? Don't you think it should be a different pull request?

hepin1989 commented:
@nafg I just want to reach to @felixmulder .and yes,it should be in another PR.

felixmulder commented:
I'm currently in Tokyo on vacation. If you're interested. I have been working on search, you can preview it here (will create a PR next week most likely):

https://dl.dropboxusercontent.com/u/358427/scaladoc/library/index.html

felixmulder commented:
Please note that I have not investigated mobile functionality and that it is a WIP.
